docs/design/features/StringDeduplication.md

```diff
@@ -0,0 +1,196 @@
+
+#### **Design outline**
+
+This is an opt in feature. When the runtime detects it’s turned on, it creates a dedup thread to do the work. 
+
+Detection of duplicated strings is done by looking into a hash table. The key into this hash table is the hash code of the content of a string. Detailed description of this detection is later in this doc. 
+
+As the dedup thread goes through the old generations linearly, it looks for references to a string object (denoted by the method table) and either calculates or looks up the hash code of that string to see if it already exists in the hash table. If so it will attempt to change the reference to point to that string with a CAS operation. If this fails, which means some other thread changed the reference at the mean time, we simply ignore this and move on. We expect the CAS failure rate to be very low.
+
+Since the new string reference we will write to the heap has the exact same type info, it means BGC can read either the old or the new reference and will get the same size info so deduping would not conflict with BGC concurrent mark. It does not conflict with BGC sweep because BGC sweep does not read references inside an object.
+
+The dedup hash table acts as weak references to the strings. Depending on the scenario we might choose to null out these weak references or not (if it’s more performant to rebuild the hash table). If we we do the former these weak references would be treated as short weak handles so the following will happen before we scan for finalization -
+
+- During BGC final mark phase we will need to null out the strings that are not marked in the hash table. This can be made concurrent. 
+
+- During a full blocking GC we will need to null out the strings that are not marked in the hash table, and relocate the ones that got promoted if we are doing a compacting GC.
+
+**Alternate design points**
+
+- Should we create multiple threads to do the work? 
+
+Deduping can be done leisurely, so it doesn’t merit having multiple threads. 
+
+- Can we use an existing thread to do the work on? 
+
+The finalizer thread is something that’s idling most of the time. However there are already plenty of types of work scheduled to potentially run on the finalizer thread so adding yet another thing, especially an opt in feature, can get messy.
+
+#### **Which strings to dedup**
+
+Only strings allocated on the managed heap will be considered for deduplication. And only the ones that live in the old generations, meaning gen2 an LOH, will be considered. Deduping in young generations would be unproductive as they will likely die very soon. It would be extremely unproductive if the copy we are deduping to actually lived in young gens.
+
+#### **Calculating hash codes and large string consideration**
+
```

Comment:
In Roslyn we found diminishing returns in deduping of large strings. In extreme case 1Gb string is expensive to hash and is unlikely to have many copies. If I remember correctly we just limited the length to be 50 or so, based on hit rates in the deduping table when compiling various projects.

Reply:
>  If I remember correctly we just limited the length to be 50 or so

While there are [over 23,404,126 user agent](https://developers.whatismybrowser.com/useragents/explore/) strings; the most comment are around 120 - 218 chars; so 250 might be a good number here? (Link offers a download of all 23M if you want to go deeper on the occurances and lengths...)

Reply:
Ah, doesn't look like the download provides occurrence counts 😢

Reply:
50 was good for c#. More general case may need a larger limit, but it feels there need to be _some_ limit.

Reply:
it's a good thing to mention that there will be a limit. obviously, this limit will be way less than 1Gb. 1Gb is simply an impractical example - I mean you can't even create many of them to begin with.